### DISCUSSION TITLE
How does Global Ray look for framework specific Ray installs?

### DISCUSSION THREAD

**Discussion by jeffreyvr** [spatie/global-ray#22] [Q&A]
I have Global Ray installed and a Laravel project running on Valet. Because I wan't to use Laravel specific features of Ray, I installed `spatie/laravel-ray` in the project.

However, when I call `ray` functions it will still use Global Ray instead of the project's version. Calling `showQueries` for example, then results in a `Method showQueries does not exist.`.

It appears to be looking in the Valet composer packages for other versions of ray, instead of in the project I'm working on.

`$composerJson` (in the `global-ray-loader.php` file) returns this path `/Users/myusername/.composer/vendor/laravel/valet/composer.json`

What is the intended behaviour?

**Comment by freekmurze** [comment]
That's strange. We check if the laravel package is installed here: https://github.com/spatie/global-ray/blob/main/src/scripts/global-ray-loader.php#L51-L55

**Comment by jeffreyvr** [comment]
Thanks Freek! So should https://github.com/spatie/global-ray/blob/main/src/scripts/global-ray-loader.php#L15 (`$composerJson`) normally return the path of the project `composer.json`?

**Reply by stevebauman** [reply]
Hey @jeffreyvr, it definitely should. How are you executing the PHP script containing the `ray()` call? CLI or web request?

**Reply by jeffreyvr** [reply]
Hi @stevebauman, Through a web request. Specifically it's in a controller method.

**Reply by stevebauman** [reply]
Nevermind -- I'm able to reproduce this on Valet by editing the `global-ray-loader.php` file and then visiting a Valet parked site:

**global-ray-loader.php**

```php
class PharLoader
{
    public static function load(string $pharPath, array $unlessDetectedPackages)
    {
        $composerJson = getcwd() . '/composer.json';

        var_dump($composerJson);
        die();

        // ...
    }
}
```

**Result**:

![Screen Shot 2022-06-21 at 9 35 11 AM](https://user-images.githubusercontent.com/6421846/174812527-1452efc8-071a-4096-b772-16661a489639.png)

**Reply by jeffreyvr** [reply]
Could do something like this? Not beautiful, but seems to work.

```php
public static function getComposerPath()
{
    $composerJson = getcwd() . '/composer.json';

    // Check if it is Laravel Valet.
    if(strpos($composerJson, 'valet') != false) {
        $valetConfig = json_decode(file_get_contents($_SERVER['HOME'].'/.config/valet/config.json'));

        foreach($valetConfig->paths as $path) {
            $composerPath = $path . '/' . str_replace('.' . $valetConfig->tld, '/', $_SERVER['HTTP_HOST']) . 'composer.json';

            if(file_exists($composerPath)) {
                return $composerPath;
            }
        }
    }

    return $composerJson;
}
```

**Reply by stevebauman** [reply]
Nice @jeffreyvr! I think that'll do just fine. Do you want to make a PR? I can create one if you'd like 👍

**Reply by jeffreyvr** [reply]
@stevebauman Done! :)

**Reply by stevebauman** [reply]
Excellent thanks @jeffreyvr! 🙏
